src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -802,6 +817,30 @@ public void Close()
                 _terminateDelegate(_nodeId);
             }
 
+            /// <summary>
+            /// Waits for the child node process to exit.
+            /// </summary>
+            public void WaitForExit()
+            {
+                int processId = _processId;
+                if (processId != -1)
+                {
+                    Process childProcess;
+                    try
+                    {
+                        childProcess = Process.GetProcessById(processId);
+                    }
+                    catch (System.ArgumentException)
+                    {
+                        // The process has terminated already.
+                        return;
+                    }
+                    // Wait for the process to terminate.
+                    CommunicationsUtilities.Trace("Waiting for node with pid = {0} to terminate", processId);
+                    childProcess.WaitForExit();
```

Comment:
Just because there can be unanticipated errors, I'd feel much more comfortable if this had a timeout.

Reply:
I've added a timeout. Followed by a Kill + blocking wait.